sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala

```diff
@@ -306,9 +296,13 @@ class SimpleFunctionRegistry
     extends SimpleFunctionRegistryBase[Expression]
     with FunctionRegistry {
 
-  override def clone(): SimpleFunctionRegistry = synchronized {
+  override def clone(): SimpleFunctionRegistry = {
     val registry = new SimpleFunctionRegistry
-    functionBuilders.iterator.foreach { case (name, (info, builder)) =>
+    val iterator = functionBuilders.entrySet().iterator()
+    while (iterator.hasNext) {
+      val entry = iterator.next()
+      val name = entry.getKey
+      val (info, builder) = entry.getValue
```

Comment:
Can we directly use `functionBuilders.forEach { (name, (info, builder))  ... ` here?